documentation/proposals/Proposal - Generic Math.md

```diff
@@ -0,0 +1,1189 @@
+
+    public TScalar LengthSquared { get; }
+
+    public Vector2I<TScalar> Dot(Vector2I<TScalar> other);
+    public TScalar DistanceSquared(Vector2I<TScalar> other);
+    public Vector2I<TScalar> Abs();
+    public void CopyTo(TScalar[] array);
+    public void CopyTo(TScalar[] array, int index);
+
+    public static Vector2I<TScalar> Dot (Vector2I<TScalar> left, Vector2I<TScalar> right);
+    public static Vector2I<TScalar> DistanceSquared (Vector2I<TScalar> left, Vector2I<TScalar> right);
+    public static Vector2I<TScalar> Clamp (Vector2I<TScalar> value, Vector2I<TScalar> min, Vector2I<TScalar> max);
+    public static Vector2I<TScalar> Min (Vector2I<TScalar> value1, Vector2I<TScalar> value2);
+    public static Vector2I<TScalar> Max (Vector2I<TScalar> value1, Vector2I<TScalar> value2);
+    public static Vector2I<TScalar> Abs (Vector2I<TScalar> value);
+}
+```
+
+```cs
+public readonly record struct Vector2F<TScalar> : IEquatable<Vector2F<TScalar>>, 
+                           IEquatable<Vector2>,
+                           IFloatingPointVector<Vector2F<TScalar>, TScalar>,
+                           IVector2<Vector2F<TScalar>, TScalar>
+                           where TScalar : IFloatingPointIeee754<TScalar>
+{
+    ///<summary>
+    ///The X component of the Vector
+    ///</summary>
+    [DataMember]
+    public TScalar X { get; }
+
+    ///<summary>
+    ///The Y component of the Vector
+    ///</summary>
+    [DataMember]
+    public TScalar Y { get; }
+
+    public static Vector2F<TScalar> UnitX { get; }
+    public static Vector2F<TScalar> UnitY { get; }
+
+    public static Vector2F<TScalar> Zero { get; }
+    public static Vector2F<TScalar> One { get; }
+
+    public TScalar this[int index] { get; }
+
+    public Vector2F(TScalar x, TScalar y);
+    public Vector2F(TScalar value);
+    public Vector2F(ReadOnlySpan<TScalar> span);
+
+    public Span<TScalar> AsSpan();
+
+    public static Vector2F<TScalar> operator +(Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> operator -(Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> operator *(Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> operator /(Vector2F<TScalar> left, Vector2F<TScalar> right);
+    public static Vector2F<TScalar> operator %(Vector2F<TScalar> left, Vector2F<TScalar> right);
+
+    public static Vector2F<TScalar> operator *(Vector2F<TScalar> left, TScalar right);
+    public static Vector2F<TScalar> operator /(Vector2F<TScalar> left, TScalar right);
+    public static Vector2F<TScalar> operator %(Vector2F<TScalar> left, TScalar right);
+
+    public static Vector2F<TScalar> operator *(TScalar left, Vector2F<TScalar> right);
+
+    public static Vector2F<TScalar> operator +(Vector2F<TScalar> value);
+    public static Vector2F<TScalar> operator -(Vector2F<TScalar> value);
+
+    public TScalar LengthSquared { get; }
+    public TScalar Length();
```

Comment:
It seems counter intuitive for `LengthSquared` to be a property but `Length` to be a method. I'd make both properties and plan to do that for the generic ones in System.Numerics when we eventually expose them.

Reply:
Agreed